File: EasyToRememberPwd-WebApi/domain/PasswordGenerator.cs
using System.Globalization;
using EasyToRemember.Cryptography;

namespace EasyToRememberPwd_WebApi
{
    public class PasswordGenerator
    {
        private readonly bool _addNumbers;
        private readonly bool _addSymbols;
        private readonly int _minimumLength;

        private readonly CryptoRandom _randomNumGenerator = new CryptoRandom();
        private readonly char[] _symbols;

        public PasswordGenerator(int minimumLength, bool addNumbers, bool addSymbols, char[] symbols)
        {
            _minimumLength = minimumLength;

            if (_minimumLength < 8)
            {
                _minimumLength = 8;
            }

            _addNumbers = addNumbers;
            _addSymbols = addSymbols;
            _symbols = symbols;
        }

        public string Generate()
        {
            string randomNumber = string.Empty;
            if (_addNumbers)
            {
                int maxNumber = _minimumLength <= 8 ? 10 : 100;
                randomNumber = _randomNumGenerator.Next(1, maxNumber).ToString();
            }

            string symbol = string.Empty;
            if (_addSymbols)
            {
                symbol = _symbols[_randomNumGenerator.Next(0, _symbols.Length)].ToString();
            }

            (int firstWordLen, int secondWordLen) = GetWordLengths(randomNumber.Length, symbol.Length);

            string firstWord = CultureInfo.CurrentCulture.TextInfo.ToTitleCase(new Adjectives().GetRandom(firstWordLen));
            string secondWord = CultureInfo.CurrentCulture.TextInfo.ToTitleCase(new Nouns().GetRandom(firstWordLen));

            return $"{firstWord}{secondWord}{randomNumber}{symbol}";
        }


        private (int firstWordLen, int secondWordLen) GetWordLengths(int randomNumberLength, int symbolLength)
        {
            int newMinLen = _minimumLength - randomNumberLength - symbolLength;

            int randNewLen = _randomNumGenerator.Next(newMinLen, newMinLen + 2);

            int toReturn1 = randNewLen / 2;
            int toReturn2 = randNewLen - toReturn1;

            return (toReturn1, toReturn2);
        }
    }
}

File: Easy-To-Remember-Password_Extension/src/browser_action/PasswordGenerator.js
class PasswordGenerator {
  constructor(minimumLength, addNumbers, addSymbols, symbols) {
    this.minimumLength = minimumLength < 8 ? 8 : minimumLength;
    this.addNumbers = addNumbers;
    this.addSymbols = addSymbols;
    this.symbols = symbols;
    this.randomGenerator = new CryptoRandom();
  }

  generate() {
    let randomNumber = "";
    if (this.addNumbers) {
      const maxNumber = this.minimumLength <= 8 ? 10 : 100;
      randomNumber = this.randomGenerator.next(1, maxNumber).toString();
    }

    let symbol = "";
    if (this.addSymbols && this.symbols.length > 0) {
      const symbolIndex = this.randomGenerator.next(0, this.symbols.length);
      symbol = this.symbols[symbolIndex];
    }

    const wordLengths = this.getWordLengths(randomNumber.length, symbol.length);

    const firstWord = this.toTitleCase(
      new Adjectives().getRandom(wordLengths.firstWordLen)
    );
    const secondWord = this.toTitleCase(
      new Nouns().getRandom(wordLengths.secondWordLen)
    );

    return `${firstWord}${secondWord}${randomNumber}${symbol}`;
  }

  getWordLengths(randomNumberLength, symbolLength) {
    const newMinLen = this.minimumLength - randomNumberLength - symbolLength;
    const randNewLen = this.randomGenerator.next(newMinLen, newMinLen + 2);

    const firstWordLen = Math.floor(randNewLen / 2);
    const secondWordLen = randNewLen - firstWordLen;

    return {
      firstWordLen: firstWordLen,
      secondWordLen: secondWordLen,
    };
  }

  toTitleCase(str) {
    if (!str) return "";
    return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
  }
}
